src/Microsoft.IdentityModel.Tokens/Validation/Validators.TokenType.cs

```diff
@@ -57,5 +59,117 @@ public static string ValidateTokenType(string type, SecurityToken securityToken,
 
             return type;
         }
+
+        /// <summary>
+        /// Validates the type of the token.
+        /// </summary>
+        /// <param name="type">The token type or <c>null</c> if it couldn't be resolved (e.g from the 'typ' header for a JWT).</param>
+        /// <param name="securityToken">The <see cref="SecurityToken"/> that is being validated.</param>
+        /// <param name="validationParameters"><see cref="TokenValidationParameters"/> required for validation.</param>
+        /// <param name="callContext"></param>
+        /// <exception cref="ArgumentNullException">If <paramref name="validationParameters"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">If <paramref name="securityToken"/> is null.</exception>
+        /// <exception cref="SecurityTokenInvalidTypeException">If <paramref name="type"/> is null or whitespace and <see cref="TokenValidationParameters.ValidTypes"/> is not null.</exception>
+        /// <exception cref="SecurityTokenInvalidTypeException">If <paramref name="type"/> failed to match <see cref="TokenValidationParameters.ValidTypes"/>.</exception>
+        /// <remarks>An EXACT match is required. <see cref="StringComparison.Ordinal"/> (case sensitive) is used for comparing <paramref name="type"/> against <see cref="TokenValidationParameters.ValidTypes"/>.</remarks>
+#pragma warning disable CA1801 // TODO: remove pragma disable once callContext is used for logging
+        internal static TokenTypeValidationResult ValidateTokenType(string? type, SecurityToken? securityToken, TokenValidationParameters validationParameters, CallContext callContext)
+#pragma warning restore CA1801 // TODO: remove pragma disable once callContext is used for logging
+        {
+            if (securityToken == null)
+            {
+                return new TokenTypeValidationResult(
+                    type,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            LogMessages.IDX10000,
+                            LogHelper.MarkAsNonPII(nameof(securityToken))),
+                        typeof(ArgumentNullException),
+                        new StackFrame(true)));
+            }
+
+            if (validationParameters == null)
+                return new TokenTypeValidationResult(
+                    type,
+                    ValidationFailureType.NullArgument,
+                    new ExceptionDetail(
+                        new MessageDetail(
+                            LogMessages.IDX10000,
+                            LogHelper.MarkAsNonPII(nameof(validationParameters))),
+                        typeof(ArgumentNullException),
+                        new StackFrame(true)));
+
+            if (validationParameters.TypeValidator == null && (validationParameters.ValidTypes == null || !validationParameters.ValidTypes.Any()))
```

Comment:
From previous PRs, I remember we are trying to avoid using the `IEnumerable` type, particulrarly with `Any()` and `foreach` as they incur an allocation for the enumerator.
Would it make sense to extract `validationParameters.ValidTypes` as a `List<T>` and use that instead?
This point could use input from someone more familiar with .NET.

This would also apply to further uses of `ValidTypes` below.

Reply:
I think this can be left with Any() for now as IEnumerable since, that is the type that comes from the `TokenValidationParameters` however, we should consider this for the new `ValidationParameters` as changing it into IList doesn't seem like a good fit to me for the small gain in allocation as it stands right now. ValidTypes in this case will only check if the ValidTypes come as empty and the other instance where we use it, is to check whether or not a type is a part of the collection, not sure if the additional overhead of transforming to IList is worth it?

Reply:
I checked with @keegan-caruso on this for an additional opinion and we think it can be left as Any() for now.